lib/async_hooks.js

```diff
@@ -223,80 +223,67 @@ const storageHook = createHook({
 class AsyncLocalStorage {
   constructor() {
     this.kResourceStore = Symbol('kResourceStore');
-    this.enabled = false;
   }
 
   disable() {
-    if (this.enabled) {
-      this.enabled = false;
-      // If this.enabled, the instance must be in storageList
-      storageList.splice(storageList.indexOf(this), 1);
-      if (storageList.length === 0) {
-        storageHook.disable();
-      }
+    const resource = executionAsyncResource();
+    resource[this.kResourceStore] = undefined;
+
+    // If this.enabled, the instance must be in storageList
+    storageList.splice(storageList.indexOf(this), 1);
+    if (storageList.length === 0) {
+      storageHook.disable();
     }
   }
 
   // Propagate the context from a parent resource to a child one
   _propagate(resource, triggerResource) {
     const store = triggerResource[this.kResourceStore];
-    if (this.enabled) {
-      resource[this.kResourceStore] = store;
-    }
+    resource[this.kResourceStore] = store;
   }
 
   _enter(store) {
-    if (!this.enabled) {
-      this.enabled = true;
+    if (storageList.indexOf(this) === -1) {
       storageList.push(this);
       storageHook.enable();
     }
     const resource = executionAsyncResource();
     resource[this.kResourceStore] = store;
   }
 
-  _exit() {
-    const resource = executionAsyncResource();
-    if (resource) {
-      resource[this.kResourceStore] = undefined;
-    }
-  }
-
   runSyncAndReturn(store, callback, ...args) {
-    this._enter(store);
-    try {
+    const resource = new AsyncResource('AsyncLocalStorage');
+    return resource.runInAsyncScope(() => {
+      this._enter(store);
       return callback(...args);
-    } finally {
-      this._exit();
-    }
+    });
   }
 
   exitSyncAndReturn(callback, ...args) {
-    this.enabled = false;
-    try {
+    const resource = new AsyncResource('AsyncLocalStorage');
+    return resource.runInAsyncScope(() => {
+      this.disable();
```

Comment:
I have limited bandwidth and computing power so I can't run tests for now, but, since `this.disable()` removes the current AsyncLocalStorage from the list, other tracks/paths would be impacted too right?

```js
asl. exitSyncAndReturn(() => {});
process.nextTick(() => {
    // is this still in context?
});
```

Reply:
Yes, that would still be in-context as the `AsyncResource` will produce its own resource, meaning the underlying stack that is used to propagate the store through would point to a different resource when the disable is triggered than the one at the level of that nextTick.

Reply:
But the current instance is not in `storageList` anymore right?

Reply:
Ah, good point. Fixed.